h11/_headers.py

```diff
@@ -90,30 +99,57 @@ class Headers(Sequence):
 
     __slots__ = "_full_items"
 
-    def __init__(self, full_items):
+    def __init__(self, full_items: List[Tuple[bytes, bytes, bytes]]) -> None:
         self._full_items = full_items
 
-    def __bool__(self):
+    def __bool__(self) -> bool:
         return bool(self._full_items)
 
-    def __eq__(self, other):
-        return list(self) == list(other)
+    def __eq__(self, other: object) -> bool:
+        return list(self) == list(other)  # type: ignore
 
-    def __len__(self):
+    def __len__(self) -> int:
         return len(self._full_items)
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return "<Headers(%s)>" % repr(list(self))
 
-    def __getitem__(self, idx):
+    def __getitem__(self, idx: int) -> Tuple[bytes, bytes]:  # type: ignore[override]
         _, name, value = self._full_items[idx]
         return (name, value)
 
-    def raw_items(self):
+    def raw_items(self) -> List[Tuple[bytes, bytes]]:
         return [(raw_name, value) for raw_name, _, value in self._full_items]
 
 
-def normalize_and_validate(headers, _parsed=False):
+HeaderTypes = Union[
+    List[Tuple[bytes, bytes]],
+    List[Tuple[bytes, str]],
+    List[Tuple[str, bytes]],
+    List[Tuple[str, str]],
+]
+
+
+@overload
+def normalize_and_validate(headers: Headers, _parsed: Literal[True]) -> Headers:
+    ...
+
+
+@overload
+def normalize_and_validate(headers: HeaderTypes, _parsed: Literal[False]) -> Headers:
+    ...
```

Comment:
I don't understand this overloading?
What's with the `Literal[True]`, `Literal[False]`, `bool` cases?

Reply:
It allows mypy to determine that the type of `headers` changes depending on `_parsed` being True or False. See also [this](https://github.com/python-hyper/h11/pull/135#discussion_r697890517)